core/packages/contracts/contracts/utils/Bitfield.sol

```diff
@@ -48,87 +41,103 @@ library Bitfield {
+            }
+
+            Location memory loc = toLocation(index);
 
             // require randomly selected bit to be set in prior
-            if (!isSet(prior, index)) {
+            if (!isSet(prior, loc)) {
+                unchecked { ++i; }
                 continue;
             }
 
             // require a not yet set (new) bit to be set
-            if (isSet(bitfield, index)) {
+            if (isSet(bitfield, loc)) {
+                unchecked { ++i; }
                 continue;
             }
 
-            set(bitfield, index);
+            set(bitfield, loc);
 
-            found++;
+            unchecked { ++found; }
+            unchecked { ++i; }
         }
 
         return bitfield;
     }
 
     function createBitfield(uint256[] calldata bitsToSet, uint256 length)
-        public
+        internal
         pure
         returns (uint256[] memory bitfield)
     {
         // Calculate length of uint256 array based on rounding up to number of uint256 needed
-        uint256 arrayLength = (length + 255) / 256;
+        uint256 arrayLength;
+        unchecked {
+            arrayLength = (length + 255) / 256;
+        }
 
         bitfield = new uint256[](arrayLength);
 
-        for (uint256 i = 0; i < bitsToSet.length; i++) {
-            set(bitfield, bitsToSet[i]);
+        for (uint256 i = 0; i < bitsToSet.length;) {
+            set(bitfield, toLocation(bitsToSet[i]));
+            unchecked { ++i; }
         }
 
         return bitfield;
     }
 
     /**
      * @notice Calculates the number of set bits by using the hamming weight of the bitfield.
-     * The alogrithm below is implemented after https://en.wikipedia.org/wiki/Hamming_weight#Efficient_implementation.
-     * Further improvements are possible, see the article above.
+     * The alogrithm below is implemented after popcount64c at https://en.wikipedia.org/wiki/Hamming_weight#Efficient_implementation.
      */
-    function countSetBits(uint256[] memory self) public pure returns (uint256) {
-        uint256 count = 0;
-        for (uint256 i = 0; i < self.length; i++) {
-            uint256 x = self[i];
-
-            x = (x & M1) + ((x >> 1) & M1); //put count of each  2 bits into those  2 bits
-            x = (x & M2) + ((x >> 2) & M2); //put count of each  4 bits into those  4 bits
-            x = (x & M4) + ((x >> 4) & M4); //put count of each  8 bits into those  8 bits
-            x = (x & M8) + ((x >> 8) & M8); //put count of each 16 bits into those 16 bits
-            x = (x & M16) + ((x >> 16) & M16); //put count of each 32 bits into those 32 bits
-            x = (x & M32) + ((x >> 32) & M32); //put count of each 64 bits into those 64 bits
-            x = (x & M64) + ((x >> 64) & M64); //put count of each 128 bits into those 128 bits
-            x = (x & M128) + ((x >> 128) & M128); //put count of each 256 bits into those 256 bits
-            count += x;
+    function countSetBits(uint256[] memory self) internal pure returns (uint256) {
```

Comment:
Hey these changes look good, didn't know further optimizations were possible. However, I'd prefer if we make them in a separate PR, especially as we'll need to add a bunch of tests.

Reply:
Cool, will do! I was following the Wikipedia article mentioned in the comment